backend/common/repo.py

```diff
@@ -0,0 +1,233 @@
+
+        return parse.urlunparse((p_url.scheme, p_url.netloc, path, p_url.params, p_url.query, p_url.fragment,))
+
+    def get_pkg_index_raw(self) -> typing.Tuple[str, bytes]:
+        """
+        Get Packages.gz or Packages.xz or Packages content, raw.
+
+        :return: bytes of the content
+        """
+        if self._pkg_index[0] == "":
+            for cnt_fname in [DpkgRepo.PKG_GZ, DpkgRepo.PKG_XZ, DpkgRepo.PKG_RW]:
+                packages_url = self.append_index_file(cnt_fname)
+                resp = requests.get(packages_url)
+                if resp.status_code == http.HTTPStatus.OK:
+                    self._pkg_index = cnt_fname, resp.content
+                    break
+                resp.close()
+
+        if self._pkg_index == ("", b"",):
+            raise GeneralRepoException("No variants of package index has been found on {} repo".format(self._url))
+
+        return self._pkg_index
+
+    def decompress_pkg_index(self) -> str:
+        """
+        Find and return contents of Packages.gz file.
+
+        :raises GeneralRepoException if the Packages.gz file cannot be found.
+        :return: string
+        """
+        fname, cnt_data = self.get_pkg_index_raw()
+        try:
+            if fname == DpkgRepo.PKG_GZ:
+                cnt_data = zlib.decompress(cnt_data, 0x10 + zlib.MAX_WBITS)
+            elif fname == DpkgRepo.PKG_XZ:
+                cnt_data = lzma.decompress(cnt_data)
+        except (zlib.error, lzma.LZMAError) as exc:
+            raise GeneralRepoException(exc)
+        except Exception as exc:
+            raise GeneralRepoException("Unhandled exception occurred while decompressing {}: {}".format(fname, exc))
+
+        return cnt_data.decode("utf-8")
+
+    def _parse_release_index(self, release: str) -> "EntryDict":
+        """
+        Parse release index to a structure.
+
+        :param release: decoded content of the Release file
+        :return: dictionary
+        """
+        for line in release.split(os.linesep):
+            cs_s_path = tuple(filter(None, line.strip().replace("\t", " ").split(" ")))
+            if len(cs_s_path) == 3 and len(cs_s_path[0]) in [0x20, 0x28, 0x40]:
+                try:
+                    int(cs_s_path[0], 0x10)
+                    rel_entry = DpkgRepo.ReleaseEntry(int(cs_s_path[1]), cs_s_path[2])
+                    self._release.setdefault(rel_entry.uri, rel_entry)
+                    if len(cs_s_path[0]) == 0x20:
+                        self._release[rel_entry.uri].checksum.md5 = cs_s_path[0]
+                    elif len(cs_s_path[0]) == 0x28:
+                        self._release[rel_entry.uri].checksum.sha1 = cs_s_path[0]
+                    elif len(cs_s_path[0]) == 0x40:
+                        self._release[rel_entry.uri].checksum.sha256 = cs_s_path[0]
+
+                except ValueError:
+                    pass
+
+        return self._release
+
+    def get_release_index(self) -> typing.Dict[str, "DpkgRepo.ReleaseEntry"]:
```

Comment:
Does this compile with old python version? Even python 2.6? (We do not run it there, but we have to compile it there)

Reply:
@mcalmer we decided to backport that in 2.x code separately without this, so it will peacefully die together with the old versions and thus we won't be carrying around obsolete paradigms anymore.

Reply:
This package is a "client tools package" as one subpackage (spacewalk-backend-libs) is needed on the clients. These clients are still SLE11 and RHEL6 which have python 2.6 and also SLES12 with python 2.7.

The packages must still "build" with these versions. Not sure if this syntax break the compile or not.

Reply:
@mcalmer if I understand correctly, [this issue](https://github.com/SUSE/spacewalk/issues/9475) is exactly a blocker at the moment. So then we should rearrange our plan with @paususe perhaps.

Reply:
for uyuni yes. But we need a backport of this at least to 4.0 and there we need to remove the not working prototypes. But uyuni can only be merged when we agreed on a solution for this.